# NLP Lab : Modèles de langue

Dans ce tp, nous allons constuire les briques principales du modèle GPT2 et entrainer un petit modèle sur des poèmes de Victor Hugo. 

Les questions sont posées dans ce notebook, puis pour executer l'entrainement, il faudra modifier le ficher `gpt_single_head.py` aussi disponible dans le reposository git.


## Données

Les données d'entrainement sont un recueil de poèmes de Victor Hugo issu du site [gutenberg.org](https://www.gutenberg.org/). Elles sont disponibles dans le répertoire `data`.

Afin de réduire la complexité du modèle, nous allons modéliser le texte au niveau caractère. Les modèles de language modélisent généralement des séquences de sous-mots en utilisant des [tokenizers](https://huggingface.co/docs/transformers/tokenizer_summary) (BPE, SentencePiece, WordPiece)

Questions :
>* En utilisant [collections.Counter](https://docs.python.org/3/library/collections.html#collections.Counter), afficher le nombre de caractères différents dans le texte et la fréquence de chaque caractère.

In [21]:
working_directory="/home/onyxia/work/nlp-lab-language-models"

In [2]:
import numpy as np

In [3]:
import collections

with open('data/hugo_contemplations.txt', 'r', encoding='utf-8') as f:
    text = f.read()

print(f'Number of characters in the file: {len(text)}')
##  YOUR CODE HERE
counter=collections.Counter(text)
###

print (f'Number of character in counter: {sum(counter.values())}')
chars=counter.keys()
print (f'{len(chars)} different characters')
print (counter)


Number of characters in the file: 285222
Number of character in counter: 285222
101 different characters
Counter({' ': 49127, 'e': 30253, 's': 17987, 'u': 14254, 'r': 14223, 't': 14071, 'a': 14048, 'n': 13725, 'i': 12828, 'o': 12653, 'l': 11638, '\n': 8102, 'm': 6495, 'd': 6375, ',': 6077, 'c': 5074, 'p': 4206, "'": 3820, 'v': 3492, 'é': 2943, 'b': 2783, 'f': 2772, 'h': 2221, 'q': 1956, 'g': 1790, '.': 1420, 'x': 1154, 'L': 1147, '!': 1121, 'E': 1074, ';': 1043, '-': 1020, 'j': 890, 'D': 764, 'è': 725, 'à': 706, 'y': 660, 'I': 627, 'ê': 605, 'C': 593, 'S': 545, 'A': 530, 'Q': 503, 'z': 482, 'J': 471, 'O': 450, 'T': 441, 'P': 435, '?': 388, 'V': 383, 'â': 381, 'N': 362, 'M': 344, 'ù': 298, ':': 294, 'R': 240, 'î': 214, 'U': 208, 'ô': 159, 'X': 150, '1': 146, 'H': 116, 'F': 114, '5': 111, '8': 93, 'B': 78, '«': 74, 'É': 70, '»': 69, 'G': 67, '4': 64, 'û': 62, '3': 47, 'ç': 34, 'À': 33, 'ë': 32, 'ï': 31, '2': 30, '·': 26, 'Ê': 24, '6': 23, '7': 23, 'Ô': 19, '9': 19, 'È': 11, 'k': 10, '0':

In [3]:
petit_msg="Bonjour tout le monde"
compteur_dict=collections.Counter(petit_msg)

In [4]:
compteur_dict.keys()

dict_keys(['B', 'o', 'n', 'j', 'u', 'r', ' ', 't', 'l', 'e', 'm', 'd'])

### Encodage / décodage
Afin de transformer le texte en vecteur pour le réseau de neurones, il faut encoder chaque caractère avec un entier. Les fonctions suivantes opérent l'encodage et le décodage des caractères :

In [4]:
# create a mapping from characters to integers
stoi = { ch:i for i,ch in enumerate(chars) }
itos = { i:ch for i,ch in enumerate(chars) }
encode = lambda s: [stoi[c] for c in s] # encoder: transform a string into a list of integers
decode = lambda l: ''.join([itos[i] for i in l]) # decoder: transform a list of integers into a string


# test that your encoder/decoder is coherent
testString = "\nDemain, dès l'aube"
assert decode(encode (testString)) ==  testString

In [17]:
decode(encode (testString))==testString

True

### Découpage Train/Validation

L'objectif étant de prédire des poèmes, il ne faut pas mélanger les lignes aléatoirement. Il faut garder l'ordre des lignes dans le texte et uniquement prendre les premiers 90% pour entrainer et les 10% restants pour contrôler l'apprentissage. 

Questions :
> * Découper en `train_data` (90%) et `val_data` (10%) en utilisant du slicing sur data.

In [5]:
import torch
# Train and validation splits
data = torch.tensor(encode(text), dtype=torch.long)
## YOUR CODE HERE
data_size=data.shape[0]
train_size=int(data_size*90/100)
train_data=data[:train_size]
test_data=data[train_size:]

# first 90% will be train, rest val

###

In [12]:
type(data),data.shape,data.shape[0],data[1:15]

(torch.Tensor,
 torch.Size([285222]),
 285222,
 tensor([ 1,  2,  3,  4,  5,  6,  7,  1,  8,  9,  0, 10,  7, 11]))

### Contexte

Le modèle de langue possède comme paramètre la taille maximale du contexte à considérer pour faire la prédiction du prochain caractère. Ce contexte est appelé `block_size`. Les données d'apprentissage sont donc des séquences de caractères consécutifs, issues de l'ensemble d'entraînenement tirées aléatoirement et de longueur `block_size`.

Si le caractère de début de la séquence est `i`, la séquence de contexte est donc :
``` x = data[i:i+block_size]```
et la valeur à prédire à chaque position dans le contexte est le caractère suivant :
```y = [data[i+1:i+block_size+1]```.



In [6]:
block_size = 8

i  = torch.randint(len(data) - block_size, (1,))
print (i)
x = train_data[i:i+block_size]
y = train_data[i+1:i+1+block_size]

for t in range(block_size):
    context = x[:t+1]
    target = y[t]
    print (f'context is >{decode(context.tolist())}< target is >{decode([target.tolist()])}<')

tensor([145612])
context is > < target is >a<
context is > a< target is >u<
context is > au< target is > <
context is > au < target is >t<
context is > au t< target is >o<
context is > au to< target is >i<
context is > au toi< target is >t<
context is > au toit< target is > <


### Définition des batchs

Les batchs d'entrainement sont constitués de plusieurs séquences de caractères tirées aléatoirement dans `train_data`. Pour choisir aléatoirement une séquence à mettre dans le batch, il faut tirer aléatoirement un point de départ dans `train_data` et extraire les `block_size` caractères suivants. Lors du tirage du point de départ, faire attention à laisser suffisamment de caractères après le point de départ pour avoir une séquence de `block_size` caractères.

Questions :
> * Créer les batchs `x` en tirant `batch_size` séquences de longeur `block_size` à  partir d'un point `i` tiré aléatoirement. Empiler les exemples avec `torch.stack`.
> * Créer les batchs `y` en ajoutant le caractère suivant la séquence `x`. Empiler les exemples avec `torch.stack`.




In [7]:
batch_size = 4
torch.manual_seed(2023)
# data loading
def get_batch(split):
    # generate a small batch of data of inputs x and targets y
    data = train_data if split == 'train' else test_data
    ## YOUR CODE HERE
    # select batch_size starting points in the data, store them in a list called 
    n=len(data)-block_size
    
    starting_points=points = torch.randperm(n)[:batch_size]


    # x is the sequence of integer starting at each straing point and of length block_size
    x=torch.stack(tuple([train_data[i:i+block_size] for i in starting_points]),0)

    # y is the character after each starting position
    y=torch.stack(tuple([train_data[i+1:i+1+block_size] for i in starting_points]),0)

    ### 
    # send data and target to device
    #device = 'cuda'
    #x, y = x.to(device), y.to(device)
    return x, y

In [8]:
x,y=get_batch("train")

In [22]:
print(x)
print("\n")
print(y)

tensor([[30, 29, 46, 35, 44, 43, 35, 53],
        [51, 14, 35, 43, 38, 22, 50, 35],
        [43, 29, 30, 29, 43, 38, 44, 35],
        [14, 53, 20, 45, 30, 39, 44, 34]])


tensor([[29, 46, 35, 44, 43, 35, 53, 21],
        [14, 35, 43, 38, 22, 50, 35, 14],
        [29, 30, 29, 43, 38, 44, 35, 14],
        [53, 20, 45, 30, 39, 44, 34, 35]])


In [9]:
x_np=x.numpy()
decode(list(x_np[2]))
encode(decode(list(x_np[2]))),x_np[2]

([43, 14, 38, 39, 34, 55, 35, 43], array([43, 14, 38, 39, 34, 55, 35, 43]))

### Premier modèle : un bigramme 

Le premier modèle que nous allons implémenter est un modèle bigramme. Il prédit le caractère suivant uniquement en fonction du caractère courant. Il est possible de stocker ce modèle dans une simple matrice : pour chaque caractère (en ligne), on stocke la distribution de probabilités sur l'ensemble des caractères suivants (en colonne). On peut donc le stocker dans une simple couche [`Embedding`](https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html).

Questions :
> * Dans le constructeur, définir une couche Embedding de taille `vocab_size` par `vocab_size`.
> * Dans forward, appliquer la couche d'embedding au batch de idx (`x`).
> * Dans forward, définir la loss comme la `cross_entropy` entre la prédiction et target (`y`).



In [10]:
import torch.nn as nn

# use a gpu if we have one
#device = 'cuda' if torch.cuda.is_available() else 'cpu'

# Simple bigram model
class BigramLanguageModel(nn.Module):

    def __init__(self, vocab_size):
        super().__init__()
        # we use a simple vocab_size times vocab_size tensor to store the probabilities 
        # of each token given a single token as context in nn.Embedding
        self.vocab_size = vocab_size
        # YOUR CODE HERE
        self.embedding_layer = nn.Embedding(vocab_size, vocab_size)
        self.linear = nn.Linear(vocab_size, vocab_size)


    def forward(self, idx, targets=None):
        # idx and targets are both (Batch,Time) tensor of integers
        # YOUR CODE HERE
        idx = idx.long()

        
        embeddings = self.embedding_layer(idx)
        #-----                           ----------------
        logits = self.linear(embeddings)
        log_probs = nn.functional.log_softmax(logits, dim=-1),nn.functional.log_softmax(logits, dim=-1)
        
        #---------------------------------------------------------------------------


        # don't compute loss if we don't have targets
        if targets is None:
            loss = None
        else:
            # change the shape of the logits and target to match what is needed for CrossEntropyLoss
            # https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html
            #Batch, Time, Channels = logits.shape
            #logits = logits.view(Batch*Time, Channels)
            #targets = targets.view(Batch*Time)
            #-----------------------------------------------
            log_probs = log_probs[0]

            Batch, Time, Channels = log_probs.shape
            log_probs = log_probs.view(Batch*Time, Channels)
            targets = targets.view(Batch*Time)



            #---------------------
        
            # negative log likelihood between prediction and target
            # YOUR CODE HERE
            loss = nn.CrossEntropyLoss()(log_probs, targets)


        return log_probs, loss

    def generate(self, idx, max_new_tokens):
        # idx is (B, T) array of indices in the current context
        for _ in range(max_new_tokens):
            # get the predictions
            logits, _ = self.forward(idx)
            # focus only on the last time step
            logits = logits[0]

            logits = logits[:, -1, :] # becomes (B, C)
            # apply softmax to get probabilities
            probs = nn.functional.softmax(logits, dim=-1) # (B, C)
            # sample from the distribution
            idx_next = torch.multinomial(probs, num_samples=1) # (B, 1)
            # append sampled index to the running sequence
            idx = torch.cat((idx, idx_next), dim=1) # (B, T+1)
        return idx

vocab_size=len(chars)
model = BigramLanguageModel(vocab_size)
# send the model to device
#m = model.to(device)


In [36]:
type(model)

__main__.BigramLanguageModel

#### Modèle avant entraînement
Le modèle n'a pas encore été entrainé, il est juste initialisé, mais on peut calculer la loss sur un batch aléatoire. Les poids étant initialisés avec une distribution normale N(0,1) sur chaque dimension, la loss attendue après l'initialisation devrait être proche de `-ln(1/vocab_size)` (l'entropie est maximale).

In [11]:
import math
xb, yb = get_batch('train')
logits, loss = model.forward(xb, yb)
print (logits.shape)
print (f'loss attendue {-math.log(1.0/vocab_size)}')
print (f'loss calculée {loss}')

torch.Size([32, 101])
loss attendue 4.61512051684126
loss calculée 5.065275192260742


Pour utiliser le modèle en prédiction, il faut lui fournir un premier caractère pour amorcer la séquence : c'est le prompt. Dans notre cas, on peut initialiser la génération avec le caractère de retour à la ligne pour débuter une nouvelle phrase.

Questions :
> * Créer un prompt avec un tenseur de taille (1,1) contenant l'entier correspondant au caractère `\n`.
> * Générer une séquence de caractères de taille 100 à partir de ce prompt avec les fonctions `m.generate` et `decode`.
> * Comment est la phrase générée ?

In [12]:
print (encode(['\n']))
## YOUR CODE HERE
prompt=torch.Tensor(encode(['\n'])).reshape(1,1)
generated_seq=model.generate(idx=prompt,max_new_tokens=100)
list_generated_seq=list(np.array(list(generated_seq[0]),dtype='int'))
decoded_generated_seq=decode(list_generated_seq)

print("voila à quoi ressemble la phrase predicte","\n",decoded_generated_seq)
###

[3]
voila à quoi ressemble la phrase predicte 
 
t8kYgèaôû-»ROXRÈIç?bgsLKX0·9Tt6MËâkGVxZtt)ârêx2ÆkÆTRy06z·-Éik9RIëK PZqjPzNn0 ÂÊéêé6mYî)(4eq5_lnÔ2ôRD


### Entrainement

Pour l'entrainement, nous utilisons un optimiseur [AdamW](https://pytorch.org/docs/stable/generated/torch.optim.AdamW.html) avec un learning rate de 1e-3. Une itération d'apprentissage consiste en 
- générer un batch
- appliquer le réseau de neurones (forward) et calculer la loss (`model(xb, yb)`)
- calculer le gradient (après avoir remis à zero le gradient cumulé) (`loss.backward()`)
- mettre à jour les paramètres (`optimizer.step()`).

In [13]:
max_iters = 100
batch_size = 4
eval_interval = 10
learning_rate = 1e-3
eval_iters = 20

@torch.no_grad() # no gradient is computed here
def estimate_loss():
    """ Estimate the loss on eval_iters batch of train and val sets."""
    out = {}
    model.eval()
    for split in ['train', 'val']:
        losses = torch.zeros(eval_iters)
        for k in range(eval_iters):
            X, Y = get_batch(split)
            logits, loss = model(X, Y)
            losses[k] = loss.item()
        out[split] = losses.mean()
    model.train()
    return out

# re-create the model
model = BigramLanguageModel(vocab_size)
#m = model.to(device)
m=model

# create a PyTorch optimizer
optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)

for iter in range(max_iters):

    # every once in a while evaluate the loss on train and val sets
    if iter % eval_interval == 0:
        losses = estimate_loss()
        print(f"step {iter}: train loss {losses['train']:.4f}, val loss {losses['val']:.4f}")

    # sample a batch of data
    xb, yb = get_batch('train')

    # evaluate the loss
    logits, loss = m(xb, yb)
    optimizer.zero_grad(set_to_none=True)
    loss.backward()
    optimizer.step()


step 0: train loss 4.7608, val loss 4.7925
step 10: train loss 4.5201, val loss 4.5879
step 20: train loss 4.2946, val loss 4.3782
step 30: train loss 4.2160, val loss 4.1515
step 40: train loss 3.9994, val loss 4.0311
step 50: train loss 3.8978, val loss 3.8757
step 60: train loss 3.8245, val loss 3.7616
step 70: train loss 3.6480, val loss 3.5745
step 80: train loss 3.5264, val loss 3.5506
step 90: train loss 3.5135, val loss 3.4485


Une fois le réseau entrainé pendant 100 itérations, on peut générer une séquence de caractères.

Questions :
> * Quel est l'effet de l'entraînement ?
> * Augmenter le nombre d'itérations à 1000 puis à 10,000, noter la loss obtenue et la phrase générée. Qu'observez-vous ?

In [14]:
idx = torch.ones((1,1), dtype=torch.long)*3
print (decode(m.generate(idx, max_new_tokens=100)[0].tolist()))


PômÈF[»lomô.EchîÊ-cû(9uag3Ck7ùîëJà»»F7UYWvU?es
N9jiUG0!?yixZX[ç4Fê?w]ja udAord
Vxbe s3cannt  lù viÔW


## Single Head Attention

Nous allons maintenant implémenter le mécanisme de base de l'attention. Pour chaque couple de mots de la séquence, ce mécanisme combine Q une *query* (l'information recherchée), K une *key* (l'information obtenue) et calcule  V une *value*, un vecteur de résultat. 

![single head attention](images/single_head_attention.png)

### Masquage
Cependant, comme nous utilisons le modèle pour générer des séquences, on ne doit pas utiliser les caractères situés après le caractère courant, car ce sont justement ces caractères que l'on cherche à prédire lors de l'apprentissage : *le futur n'est pas utilisé pour prédire (le futur).* 

On va donc intégrer une matrice de masquage dans le processus. Cette matrice indique que pour le premier caractère de la séquence, on ne peux utiliser que ce caractère pour prédire (pas de contexte). Pour le second caractère, on peut utiliser le premier caractère et le second. Pour le troisième caractère, on peut utiliser le premier, le second et le troisième et ainsi de suite. Cette matrice est donc une matrice triangulaire inférieure. Cette matrice est normalisée par ligne (les lignes somment à 1).

In [15]:
T = 8

# first version of the contraints with matrix multiplication
# create a lower triangular matrix
weights0 = torch.tril(torch.ones(T,T))
# normalize each row
weights0 = weights0 / weights0.sum(1, keepdim=True) 
print (weights0)

tensor([[1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.5000, 0.5000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.3333, 0.3333, 0.3333, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.2500, 0.2500, 0.2500, 0.2500, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.2000, 0.2000, 0.2000, 0.2000, 0.2000, 0.0000, 0.0000, 0.0000],
        [0.1667, 0.1667, 0.1667, 0.1667, 0.1667, 0.1667, 0.0000, 0.0000],
        [0.1429, 0.1429, 0.1429, 0.1429, 0.1429, 0.1429, 0.1429, 0.0000],
        [0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250]])


La couche [`softmax`](https://pytorch.org/docs/stable/generated/torch.nn.functional.softmax.html) est une autre manière de réaliser la normalisation :

Question :
> * Vérifier qu'on obtient bien la même matrice.

In [16]:
tril = torch.tril(torch.ones(T,T))
weights = torch.tril(torch.ones(T,T))
weights = weights.masked_fill(tril== 0, float('-inf'))
weights = nn.functional.softmax(weights, dim=-1)
print (weights)

tensor([[1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.5000, 0.5000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.3333, 0.3333, 0.3333, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.2500, 0.2500, 0.2500, 0.2500, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.2000, 0.2000, 0.2000, 0.2000, 0.2000, 0.0000, 0.0000, 0.0000],
        [0.1667, 0.1667, 0.1667, 0.1667, 0.1667, 0.1667, 0.0000, 0.0000],
        [0.1429, 0.1429, 0.1429, 0.1429, 0.1429, 0.1429, 0.1429, 0.0000],
        [0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250]])


In [91]:
(weights.numpy()==weights0.numpy()).sum()==weights.shape[0]*weights.shape[1]

True

### Implémentation

Nous pouvons maintenant implémenter la couche d'attention :

![attention_formula](images/attention_formula.png)

Questions :

> * Créer les couches key, query et value comme des couches linéaires de dimension `C` x `head_size`.
> * Appliquer les couches à `x`.
> * `weights = query x key` (transposer les deuxième et troisième dimensions de key pour pouvoir faire le produit).
> * Appliquer le facteur de normalisation.
> * Appliquer le masque triangulaire et la softmax à `weights`.
> * Appliquer value à `x`.
> * Le résultat `out` est la multiplication de `weights` par `value(x)`.


In [18]:
head_size = 16
B, T, C = 4, 8, 32
x = torch.randn(B, T, C)
## YOUR CODE HERE
# define the Key layer  
key = nn.Linear(C,head_size)
# define the Query layer
query = nn.Linear(C,head_size)
# define the Value layer
value =  nn.Linear(C,head_size)
# apply each layer to the input
k =key(x)  # (B, T, head_size)
q =query(x)  # (B, T, head_size)
v =value(x) # (B, T, head_size)
# compute the normalize product between Q and K 
k_t=torch.transpose(k, 1, 2)
weights =q@k_t   # (B, T, head_size) @ (B, 16, head_size) -> (B, T, T)
# apply the mask (lower triangular matrix)
weights = weights.masked_fill(tril== 0, float('-inf'))
# apply the softmax
weights =  nn.functional.softmax(weights, dim=-1)
###
out  = weights @ value(x) # (B, T, head_size)

# print the result
weights[0]
out[0]

tensor([[-1.8262e-01,  2.8143e-01, -9.5965e-01, -6.7524e-01,  1.4250e-01,
         -3.3926e-01,  1.4913e-01, -9.0075e-01,  5.0410e-01,  2.0885e-01,
         -7.2768e-02, -4.4058e-01, -7.4483e-03, -6.8030e-01, -1.1095e+00,
         -1.7671e-01],
        [ 9.4657e-04, -2.1648e-01, -3.5870e-01,  6.3169e-02, -1.5064e-01,
         -7.0589e-03,  3.2341e-01, -7.6347e-01, -3.9274e-02,  5.0347e-01,
         -4.6361e-01, -5.5028e-01, -5.9986e-01, -3.7587e-01, -9.1065e-01,
          2.4285e-02],
        [-1.2395e-01, -3.5153e-01, -5.7789e-01,  2.7425e-01, -4.2474e-01,
          9.0562e-02,  1.0888e-01, -2.6354e-01, -1.8336e-01,  4.7597e-01,
          1.9135e-01,  3.3756e-01, -5.9226e-01, -8.2034e-02, -5.3525e-01,
         -1.2761e-01],
        [ 1.0252e+00, -1.5595e+00, -3.6486e-01,  1.1320e-01,  7.1510e-01,
         -4.6707e-01, -2.0542e-02,  3.9382e-01,  6.6322e-01,  2.8201e-01,
          6.9537e-01, -4.9074e-01, -6.1460e-01, -1.1143e-01, -2.8524e-01,
         -5.8009e-01],
        [-9.8409e-02

In [99]:
v.shape,torch.transpose(v, 0, 1).shape

(torch.Size([4, 8, 16]), torch.Size([8, 4, 16]))

In [19]:
import gpt_single_head

0.009989 M parameters


step 0: train loss 4.7054, val loss 4.6731
step 500: train loss 2.7430, val loss 2.8467
step 1000: train loss 2.4996, val loss 2.5927
step 1500: train loss 2.4571, val loss 2.5604
step 2000: train loss 2.4131, val loss 2.5589
step 2500: train loss 2.3903, val loss 2.5651
step 3000: train loss 2.3752, val loss 2.5359
step 3500: train loss 2.3536, val loss 2.5113
step 4000: train loss 2.3521, val loss 2.4266
step 4500: train loss 2.3412, val loss 2.4794
step 4999: train loss 2.3423, val loss 2.4739

Arsure!
Qu'é, 4isoncoireyes, à cerice l,
Vous an;
Sinle; me ivur D''e,
 Quomen tran duite
L'e,
Sand ux évetuns qu'ét-istux curiquies!
Dosour'ens, orêmes s;
Uz t mantorimau'ins,
L'iet, lau du he  episens  pire s.


 S'ininetrosupitrane t les reu;
  jime; tandez, blabarit t-ni aitére jer;
Pau,
   l'onuborboiflquir,
Etrenrvoignessces t des!
Let melitéess l'anseu meux de mouper de s seplus s,
Donien  sounvordabinte pront: labre s lelareurspensplauetre,
Me;
Quane namitscutenvon que doyche Motre u 

Questions :

> * Copier votre code dans `gpt_single_head.py` : la définition des couches dans le constructeur de la classe `Head` et les calculs dans la fonction `forward`.
> * Faire un entrainement.
> * Quelle loss en train et val obtenez vous ? Le texte vous parait-il meilleur ?


## Multi-head attention

La *multi-head attention* est simplement le calcul en parallèle de plusieurs *single head attention*. Chacune des single head attention est concaténée pour créer la sortie de la multi-head attention. Dans la figure issue de l'article original, le nombre de *heads* dans le *multi-head* est `h`. Afin d'opérer des combinaisons pondérées sur la sortie de chacune des single head, une couche de calcul linéaire est ajoutée.

![multi head attention](images/multi_head_attention.png)

Le code ci-dessous crée un module de multi-head attention.
Questions :
> * Dans le constructeur, créer une liste contenant `num_heads` module `Head` en utilisant la fonction [ModuleList](https://pytorch.org/docs/stable/generated/torch.nn.ModuleList.html) de pytorch.
> * Dans la fonction `forward`, appliquer chaque single head à l'input et concaténer le résultat en utilisant la fonction [cat](https://pytorch.org/docs/stable/generated/torch.cat.html) de pytorch.

In [20]:
class MultiHeadAttention(nn.Module):
    """ multiple heads of self-attention in parallel """

    def __init__(self, num_heads, head_size):
        super().__init__()
        ## YOUR CODE HERE
        ## list of num_heads modules of type Head
        self.heads = nn.ModuleList([Head(head_size) for _ in range(num_heads)])
        ###
        
    def forward(self, x):
        ## YOUR CODE HERE
        ## apply each head in self.heads to x and concat the results 
        out = torch.cat([h(x) for h in self.heads], dim=-1)

        return out


Questions :
> * Copier le fichier `gpt_single_head.py` en `gpt_multi_head.py`.
> * Ajouter le module MultiHeadAttention dans `gpt_multi_head.py`.
> * En tête de fichier, ajouter un paramètre  `n_head = 4`.
> * Dans le module BigramLanguageModel, remplacer le module Head par un module MultiHeadAttention avec les paramètres `num_heads = n_head` et `head_size = n_embd // n_head` pour garder le même nombre de paramètres.
> Relancer l'entrainement et noter le nombre de paramètres et les loss obtenues.

0.009893 M parameters
step 4999: train loss 2.1570, val loss 2.1802

In [24]:
#Your code here
 ## Je relance l'entrainement -mais cette fois le modele à multiHeadAttention
import gpt_multi_head
# Relance l'entrainement aprés l'ajout de la couche FeedForward

model = gpt_multi_head.BigramLanguageModel()
#m = model.to(device)
m=model
# print the number of parameters in the model
print(sum(p.numel() for p in m.parameters())/1e6, 'M parameters')

# create a PyTorch optimizer
optimizer = torch.optim.AdamW(m.parameters(), lr=learning_rate)

for iter in range(max_iters):

    # every once in a while evaluate the loss on train and val sets
    if iter % eval_interval == 0 or iter == max_iters - 1:
        losses = estimate_loss()
        print(f"step {iter}: train loss {losses['train']:.4f}, val loss {losses['val']:.4f}")

    # sample a batch of data
    xb, yb = get_batch('train')

    # evaluate the loss
    logits, loss = m(xb, yb)
    optimizer.zero_grad(set_to_none=True)
    loss.backward()
    optimizer.step()

0.009989 M parameters
step 0: train loss 4.7054, val loss 4.6731
step 500: train loss 2.7430, val loss 2.8467
step 1000: train loss 2.4996, val loss 2.5927
step 1500: train loss 2.4571, val loss 2.5604
step 2000: train loss 2.4131, val loss 2.5589
step 2500: train loss 2.3903, val loss 2.5651
step 3000: train loss 2.3752, val loss 2.5359
step 3500: train loss 2.3536, val loss 2.5113
step 4000: train loss 2.3521, val loss 2.4266
step 4500: train loss 2.3412, val loss 2.4794
step 4999: train loss 2.3423, val loss 2.4739

Arsure!
Qu'é, 4isoncoireyes, à cerice l,
Vous an;
Sinle; me ivur D''e,
 Quomen tran duite
L'e,
Sand ux évetuns qu'ét-istux curiquies!
Dosour'ens, orêmes s;
Uz t mantorimau'ins,
L'iet, lau du he  episens  pire s.


 S'ininetrosupitrane t les reu;
  jime; tandez, blabarit t-ni aitére jer;
Pau,
   l'onuborboiflquir,
Etrenrvoignessces t des!
Let melitéess l'anseu meux de mouper de s seplus s,
Donien  sounvordabinte pront: labre s lelareurspensplauetre,
Me;
Quane namitscutenv

## Ajout d'une couche de calcul FeedForward


Après les couches d'attention qui collectent l'information dans la séquence, une couche de calcul est ajoutée pour combiner toutes les informations de la séquence. Cette couche est un simple Multi-Layer-Perceptron avec une couche cachée et une non linéarité de type [RELU](https://pytorch.org/docs/stable/generated/torch.nn.ReLU.html).
![multi feedfoward](images/multi_ff.png)

In [ ]:
class FeedForward(nn.Module):
    """ a simple MLP with RELU """

    def __init__(self, n_embd):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(n_embd, n_embd),
            nn.ReLU(),
        )

    def forward(self, x):
        return self.net(x)

Questions :
> * Ajouter le module `FeedForward` dans votre fichier `gpt_multi_head.py`.
> * Ajouter cette couche `FeedForward` après la *multi-head attention*.
> * Relancer l'entrainement et noter le nombre de paramètres et les loss obtenues.

0.010949 M parameters
step 4999: train loss 2.1290, val loss 2.1216

In [32]:
# Relance l'entrainement aprés l'ajout de la couche FeedForward

import gpt_multi_head
model = gpt_multi_head.BigramLanguageModel()
#m = model.to(device)
m=model
# print the number of parameters in the model
print(sum(p.numel() for p in m.parameters())/1e6, 'M parameters')

# create a PyTorch optimizer
optimizer = torch.optim.AdamW(m.parameters(), lr=learning_rate)

for iter in range(max_iters):

    # every once in a while evaluate the loss on train and val sets
    if iter % eval_interval == 0 or iter == max_iters - 1:
        losses = estimate_loss()
        print(f"step {iter}: train loss {losses['train']:.4f}, val loss {losses['val']:.4f}")

    # sample a batch of data
    xb, yb = get_batch('train')

    # evaluate the loss
    logits, loss = m(xb, yb)
    optimizer.zero_grad(set_to_none=True)
    loss.backward()
    optimizer.step()

0.009989 M parameters
step 0: train loss 4.6971, val loss 4.6478
step 10: train loss 4.5634, val loss 4.5371
step 20: train loss 4.4626, val loss 4.4290
step 30: train loss 4.3077, val loss 4.3220
step 40: train loss 4.1779, val loss 4.1354
step 50: train loss 3.9442, val loss 3.9470
step 60: train loss 3.6716, val loss 3.6776
step 70: train loss 3.5308, val loss 3.5383
step 80: train loss 3.3552, val loss 3.4909
step 90: train loss 3.2579, val loss 3.3428
step 99: train loss 3.2831, val loss 3.3238


## Empiler les blocs

Le réseau construit jusqu'à présent n'est en fait qu'un bloc du réseau final. Il est maintenant possible d'empiler les blocs de *multi-head attention* pour créer un réseau profond. 

![multi feedfoward](images/multi_bloc.png)


Le code suivant crée un bloc : 

In [ ]:
class Block(nn.Module):
    """ A single bloc of multi-head attention """

    def __init__(self, n_embd, n_head):
        super().__init__()
        head_size = n_embd // n_head
        self.sa = MultiHeadAttention(n_head, head_size)
        self.ffwd = FeedForward(n_embd)

    def forward(self, x):
        x = self.sa(x)
        x = self.ffwd(x)
        return x

Questions :
> * Ajouter le module `Block` dans `gpt_multi_head.py`.
> * Modifier le code de `BigramLanguageModel` pour ajouter 3 `Block(n_embd, n_head=4)` avec un container [Sequential](https://pytorch.org/docs/stable/generated/torch.nn.Sequential.html) à la place de `MultiHeadAttention`et `FeedForward`.
> * Relancer l'entrainement et noter le nombre de paramètres et les loss obtenues.

0.019205 M parameters
step 4999: train loss 2.2080, val loss 2.2213

In [35]:
# Relance l'entrainement aprés l'ajout de la couche FeedForward

import gpt_multi_head
model = gpt_multi_head.BigramLanguageModel()
#m = model.to(device)
m=model
max_iters=5000
eval_interval=500

# print the number of parameters in the model
print(sum(p.numel() for p in m.parameters())/1e6, 'M parameters')

# create a PyTorch optimizer
optimizer = torch.optim.AdamW(m.parameters(), lr=learning_rate)

for iter in range(max_iters):

    # every once in a while evaluate the loss on train and val sets
    if iter % eval_interval == 0 or iter == max_iters - 1:
        losses = estimate_loss()
        print(f"step {iter}: train loss {losses['train']:.4f}, val loss {losses['val']:.4f}")

    # sample a batch of data
    xb, yb = get_batch('train')

    # evaluate the loss
    logits, loss = m(xb, yb)
    optimizer.zero_grad(set_to_none=True)
    loss.backward()
    optimizer.step()

0.009989 M parameters
step 0: train loss 4.7494, val loss 4.7594
step 500: train loss 3.0354, val loss 3.1673
step 1000: train loss 3.0427, val loss 2.9740
step 1500: train loss 2.7739, val loss 2.7623
step 2000: train loss 2.6491, val loss 2.7073
step 2500: train loss 2.5015, val loss 2.6362
step 3000: train loss 2.6021, val loss 2.4505
step 3500: train loss 2.5978, val loss 2.5483
step 4000: train loss 2.4463, val loss 2.5227
step 4500: train loss 2.4515, val loss 2.4845
step 4999: train loss 2.5389, val loss 2.5165


In [36]:
working_directory

'/home/onyxia/work/nlp-lab-language-models'

## Amélioration de l'entraînement

Si on veut continuer à augmenter la taille du réseau, il est nécessaire d'utiliser des couches permettant d'améliorer l'entraînement et ses capacités de généralisation (réduire le sur-apprentissage). Ces couches sont :
- *skip connections* ou *residual connections*
- les couches de normalisation
- le dropout.


![multi feedfoward](images/multi_skip_norm.png)


Questions :
> * Dans le module Block, ajouter une skip connection en ajoutant l'input dans chaque connexion :
```
        x = x + self.sa(self.ln1(x))
        x = x + self.ffwd(self.ln2(x))
```
> * Dans le module Block, ajouter 2 couches de [LayerNorm](https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html) de taille `n_embd` avant la couche de `Multi-Head attention` et avant la `FeedForward`.
> * Après la série de 3 blocs, ajouter une couche de LayerNorm de taille `n_embd`.
> * Définir une variable `dropout = 0.2` en début de fichier et ajouter une couche de [Dropout](https://pytorch.org/docs/stable/generated/torch.nn.Dropout.html) :
>    * Après la couche RELU dans FeedForward
>    * Après la couche de MultiHead dans `MultiHeadAttention`
>    * Après la softmax dans la single head attention `Head`.
> * Relancer l'entrainement et noter le nombre de paramètres et les loss obtenues.

0.019653 M parameters

## Conclusion

Les principaux éléments de GPT2 sont en place, il faut maintenant faire passer le modèle à l'échelle et l'entraîner sur une base de données beaucoup plus grande. Pour comparaison, les paramètres de [GPT2](https://huggingface.co/transformers/v2.11.0/model_doc/gpt2.html) sont : 

* `vocab_size = 50257` : GPT2 modélise des tokens (subwords) alors que nous modélisons des caractères. Pour nous, `vocab_size = 100`.
* `n_positions = 1024` : la taille maximale du contexte. Pour nous, c'est `block_size = 8`.
* `n_embd = 768`:  la dimension des embeddings. Pour nous c'est `n_embd = 32`.
* `n_layer = 12`: le nombre de block. Pour nous c'est 3.
* `n_head = 12`: le nombre de multi-head attention. Pour nous c'est 4.

Au total, GPT2 est composé de 1,500 millions de paramètres et a été entrainé sur 8M de pages web, soit 40 Gb de texte.


```
10.816613 M parameters
step 0: train loss 4.7847, val loss 4.7701
step 4999: train loss 0.2683, val loss 2.1161
time: 31m47.910s   
    
Le pêcheur où l'homme en peu de Carevante
Sa conter des chosses qu'en ses yoitn!

Ils sont là-hauts parler à leurs ténèbres
A ceux qu'on rêve aux oiseaux des cheveux,
Et celus qu'on tourna jamais sous le front;
Ils se disent tu mêle aux univers.
J'ai vu Jean vu France, potte; petits contempler,
Et petié calme au milibre et versait,
M'éblouissant, emportant, écoute, ingorancessible,
On meurt s'efferayait.....--Pas cont âme parle en Apparia!
```